### PR TITLE
feat(kumacp): mesh trace docs

### PR DESCRIPTION
**Pull request by slonka**
Closes https://github.com/kumahq/kuma-website/issues/1017

Did you sign your commit? [Instructions](https://github.com/kumahq/.github/blob/main/CONTRIBUTING.md#sign-your-commits) yes

Have you read [Contributing guidelines](https://github.com/kumahq/.github/blob/main/CONTRIBUTING.md)? yes

### CODE REVIEW

File: docs/docs/dev/policies/mesh-trace.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
+# Mesh Trace (beta)
```
Comment (slonka) [inline docs/docs/dev/policies/mesh-trace.md:1]: Would love feedback on how to mark that this policy is not ready to be used & clashes with TrafficTrace

File: docs/docs/dev/policies/mesh-trace.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+```yaml
+sampling:
+  overall:
+    value: 80
+  random:
+    value: 60
+  client:
+    value: 40
+```
```
Comment (lahabana) [inline docs/docs/dev/policies/mesh-trace.md:310]: Please say which one is the most likely useful :)

File: docs/docs/dev/policies/mesh-trace.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+::: tab "Full example"
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrace
+metadata:
+  name: default
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
```
Comment (lahabana) [inline docs/docs/dev/policies/mesh-trace.md:173]:
```suggestion
    kuma.io/mesh: default # optional, defaults to `default` if unset
```

File: docs/docs/dev/policies/mesh-trace.md (hunk 1 above)
Comment (lahabana) [reply docs/docs/dev/policies/mesh-trace.md:1]: I'm fine with this as it matches Kong's policy: https://docs.konghq.com/gateway/latest/stability/#main

File: docs/docs/dev/policies/mesh-trace.md [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+    sampling:
+      overall:
+        value: 80
+      random:
+        value: 60
+      client:
+        value: 40
+```
+:::
+::: tab "Minimal example"
```
Comment (lahabana) [inline docs/docs/dev/policies/mesh-trace.md:76]: Probably have minimal first

File: docs/docs/dev/policies/mesh-trace.md [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+For `header` - if a value is missing then `default` will be used.
+If `default` is not provided then the tag won't be added.
+
+## Targeting parts of the infrastructure
+
+With a lot of traffic going through the system,
+a valid use case is to target specific set of microservices that are part of a main request path
+(e.g. a process of buying an item in an e-commerce application).
+
+You can achieve that by using more specific `targetRef` and (if needed) a separate backend.
```
Comment (lahabana) [inline docs/docs/dev/policies/mesh-trace.md:340]:
```suggestion
You can achieve that by using a more specific `targetRef` and (if needed) a separate backend.
```

File: docs/docs/dev/policies/mesh-trace.md [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+```
+
+For `header` - if a value is missing then `default` will be used.
+If `default` is not provided then the tag won't be added.
+
+## Targeting parts of the infrastructure
+
+With a lot of traffic going through the system,
+a valid use case is to target specific set of microservices that are part of a main request path
+(e.g. a process of buying an item in an e-commerce application).
```
Comment (lahabana) [inline docs/docs/dev/policies/mesh-trace.md:338]: I think this paragraph is not super clear.

I think this to make sense you really need to add an example.

File: docs/docs/dev/policies/mesh-trace.md (hunk 4 above)
Comment (slonka) [reply docs/docs/dev/policies/mesh-trace.md:76]: WDYT about this nested style? I wanted to have a way to show the minimal example (something that people can start with and there is not too much tinkering) and the full example so it's easy to figure out what's possible. Do you think it's worth having this structure as a "template" for future policies?

File: docs/docs/dev/policies/mesh-trace.md (hunk 6 above)
Comment (slonka) [reply docs/docs/dev/policies/mesh-trace.md]: changed to be the mirror of the "tip" in old docs

**Comment by lahabana** [comment]
@slonka I was thinking about something last night.
It would be nice to have a standard table or something like this at the top of each policy that sums up some of the feature for example:

|use To  targetRef | use From targetRef | Works with builtinGateway | Works with externalService |
|-------|----------| ---------------------------| ----------------------------|
| x.        | x             | x                                             | x |


This is just an idea but I feel it would help users quickly understand a policy

**Comment by slonka** [comment]
@lahabana - yeah, that's a great idea, but I'd love some clarification on a couple of  things:
- "use To targetRef" - do you mean that you can use "To" in this policy, or do you think we should list which "targets" (Mesh, MeshSubset, Service)... are available?
- is "Works with builtinGateway" equivalent to being able to target `MeshGatewayRoute`? Why is `builtinGateway` singled out here, we could do tracing on delegated gateway right? 
- "Works with externalService" - we don't have a specific target for external / internal services, it's all `MeshService` right?

I think a table like this would be very nice but I'd structure it a bit differently:

MeshTrace:

| `targetRef`            | `to.targetRef` | `from.targetRef` |
|------------------------|----------------|------------------|
| Mesh                   | x              | x                |
| MeshSubset             |                |                  |
| MeshService            |                |                  |
| MeshServiceSubset      |                |                  |
| MeshGatewayRoute       |                |                  |

---

MeshAccessLog:

| `targetRef`            | `to.targetRef` | `from.targetRef` |
|------------------------|----------------|------------------|
| Mesh                   | Mesh           | Mesh             |
| MeshSubset             | MeshService    |                  |
| MeshService            |                |                  |
| MeshServiceSubset      |                |                  |
| MeshGatewayRoute       |                |                  |
| MeshHTTPRoute          |                |                  |

WDYT?

**Comment by lahabana** [comment]
Indeed we need to indicate this:

| TargetRef type | top level | to | from   |
|---------------|----------| ---| ------ |
| Mesh | x |  | |
| MeshSubset |  | x | |


Feels like this might be a bit more readable.

**Comment by slonka** [comment]
> Feels like this might be a bit more readable.

agree! 👍 - for now I'll just create a table (since there is a migration to jekyll and creating a reusable component in vuepress might mean we have to re-do the work)

**Comment by lahabana** [comment]
Agreed let's stick with markdown for now

File: docs/docs/dev/policies/mesh-trace.md [hunk 7]
```diff
@@ -290,7 +303,7 @@ By default, this property is set to false.
 ### Sampling
 
 You can configure sampling settings equivalent to Envoy's:
-- [overall](https://www.envoyproxy.io/docs/envoy/v1.22.5/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto.html?highlight=overall_sampling#extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-tracing)
+- [overall](https://www.envoyproxy.io/docs/envoy/v1.22.5/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto.html?highlight=overall_sampling#extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-tracing) - the most commonly used
 - [random](https://www.envoyproxy.io/docs/envoy/v1.22.5/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto.html?highlight=random_sampling#extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-tracing)
 - [client](https://www.envoyproxy.io/docs/envoy/v1.22.5/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto.html?highlight=client_sampling#extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-tracing)
```
Comment (lahabana) [inline docs/docs/dev/policies/mesh-trace.md:308]: I'd go with:

```
tip::
Most of the time setting only `overall` is sufficient; `random` and `client` are for advanced use cases.
```

File: docs/docs/dev/policies/matching.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+A combination of a policy and the top level `targetRef` can influence which types are available in `to` and `from` levels
+(e.g. `MeshAccessLog` with a top level `targetRef` of `MeshGatewayRoute` can only have `from` level because there is no outbound listener).
+
+If a policy does not support a level then every type will be marked with ❌ in that level
+and the YAML configuration must omit that field from the definition.
+
+# Matching meaning
+
+Top level `targetRef` defines which set of proxies a policy will affect.
+`To` level `targetRef` defines rules for outgoing traffic relative to the top level.
```
Comment (lahabana) [inline docs/docs/dev/policies/matching.md:54]:
```suggestion
`Top` level `targetRef` defines which set of proxies a policy will affect (.i.e: the set of proxies whose configuration is getting modified).
`To` level `targetRef` defines rules that applies to outgoing traffic of proxies selected by the `Top` level `targetRef`.
`From` level `targetRef` defines rules that applies to incoming traffic of proxies selected by the `Top` level `targetRef`.
```

File: docs/docs/dev/policies/matching.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        name: default
+      default:
+        backends:
+          - file:
+              format:
+                plain: '{"start_time": "%START_TIME%"}'
+              path: '/tmp/logs.txt'
+```
+
+This policy will target a service `web-frontend`.
```
Comment (lahabana) [inline docs/docs/dev/policies/matching.md:93]:
```suggestion
This policy will target all proxies that implement the service `web-frontend`.
```

I think we should stop talking a `service` when we're talking about proxies that run a service. It's less confusing

File: docs/docs/dev/policies/matching.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+It instructs Kuma to log traffic **coming** from `web-backend` to `web-frontend`.
```
Comment (lahabana) [inline docs/docs/dev/policies/matching.md:94]:
```suggestion
It logs traffic **coming** from `web-backend` to `web-frontend`.
```

File: docs/docs/dev/policies/matching.md [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+It also instructs to log traffic going out of `web-frontend` to **anything** in the `Mesh` named `default`.
```
Comment (lahabana) [inline docs/docs/dev/policies/matching.md:95]:
```suggestion
It also logs traffic going out of proxies that implement `web-frontend` to **anything** in the `Mesh`.
```

File: docs/docs/dev/policies/matching.md [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+      default:
+        backends:
+          - file:
+              format:
+                plain: '{"start_time": "%START_TIME%"}'
+              path: '/tmp/logs.txt'
+  from:
+    - targetRef: # from level targetRef
+        kind: Mesh
+        name: default
```
Comment (lahabana) [inline docs/docs/dev/policies/matching.md:84]: Let's not put `name` in Mesh we don't do XMesh stuff and it's just confusing user atm. Actually adding validation to disable setting it

File: docs/docs/dev/policies/mesh-trace.md [hunk 13]
```diff
@@ -29,6 +29,19 @@ For HTTP you can also manually forward the following headers:
...
+| TargetRef type    | top level | to  | from |
+|-------------------|-----------|-----|------|
+| Mesh              | ✅         | ❌   | ❌    |
+| MeshSubset        | ✅         | ❌   | ❌    |
+| MeshService       | ✅         | ❌   | ❌    |
+| MeshServiceSubset | ✅         | ❌   | ❌    |
+| MeshGatewayRoute  | ✅         | ❌   | ❌    |
+| MeshHTTPRoute     | ❌         | ❌   | ❌    |
+
+[Legend](matching.md#matching-matrix)
```
Comment (lahabana) [inline docs/docs/dev/policies/mesh-trace.md:43]:
```suggestion
If you don't understand this table you should read [matching docs](matching.md#matching-matrix)
```

**Review by lahabana — COMMENTED** [review]
Good additions! I think your new page will need to be extended with extra info to help users get started with the new policy but that can be done later

File: docs/docs/dev/policies/mesh-trace.md (hunk 7 above)
Comment (slonka) [reply docs/docs/dev/policies/mesh-trace.md:312]: added

File: docs/docs/dev/policies/matching.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+A policy might not support every level
+(e.g. `MeshTrace` only allows top level `targetRef`).
+
+A combination of a policy and the top level `targetRef` can influence which types are available in `to` and `from` levels
+(e.g. `MeshAccessLog` with a top level `targetRef` of `MeshGatewayRoute` can only have `from` level because there is no outbound listener).
+
+If a policy does not support a level then every type will be marked with ❌ in that level
+and the YAML configuration must omit that field from the definition.
+
+# targetRef levels
```
Comment (lahabana) [inline docs/docs/dev/policies/matching.md:51]:
```suggestion
## targetRef levels
```

Also I think this section should be before the `matrix` one

**Review by lahabana — APPROVED** [review]

**Pull request merged**